python-sdk/nuscenes/eval/tracking/evaluate.py

```diff
@@ -0,0 +1,252 @@
+        self.cfg = config
+        self.result_path = result_path
+        self.eval_set = eval_set
+        self.output_dir = output_dir
+        self.verbose = verbose
+
+        # Check result file exists.
+        assert os.path.exists(result_path), 'Error: The result file does not exist!'
+
+        # Make dirs.
+        self.plot_dir = os.path.join(self.output_dir, 'plots')
+        if not os.path.isdir(self.output_dir):
+            os.makedirs(self.output_dir)
+        if not os.path.isdir(self.plot_dir):
+            os.makedirs(self.plot_dir)
+
+        # Load data.
+        if verbose:
+            print('Initializing nuScenes tracking evaluation')
+        pred_boxes, self.meta = load_prediction(self.result_path, self.cfg.max_boxes_per_sample, TrackingBox,
+                                                     verbose=verbose)
+        gt_boxes = load_gt(nusc, self.eval_set, TrackingBox, verbose=verbose)
+
+        assert set(pred_boxes.sample_tokens) == set(gt_boxes.sample_tokens), \
+            "Samples in split don't match samples in predicted tracks."
+
+        # Add center distances.
+        pred_boxes = add_center_dist(nusc, pred_boxes)
+        gt_boxes = add_center_dist(nusc, gt_boxes)
+
+        # Filter boxes (distance, points per box, etc.).
+        if verbose:
+            print('Filtering tracks')
+        pred_boxes = filter_eval_boxes(nusc, pred_boxes, self.cfg.class_range, verbose=verbose)
+        if verbose:
+            print('Filtering ground truth tracks')
+        gt_boxes = filter_eval_boxes(nusc, gt_boxes, self.cfg.class_range, verbose=verbose)
+
+        self.sample_tokens = gt_boxes.sample_tokens
+
+        # Convert boxes to tracks format.
+        self.tracks_gt = create_tracks(gt_boxes, nusc, self.eval_set, gt=True)
+        self.tracks_pred = create_tracks(pred_boxes, nusc, self.eval_set, gt=False)
+
+    def evaluate(self) -> Tuple[TrackingMetrics, TrackingMetricDataList]:
+        """
+        Performs the actual evaluation.
+        :return: A tuple of high-level and the raw metric data.
+        """
+        start_time = time.time()
+        metrics = TrackingMetrics(self.cfg)
+
+        # -----------------------------------
+        # Step 1: Accumulate metric data for all classes and distance thresholds.
+        # -----------------------------------
+        if self.verbose:
+            print('Accumulating metric data...')
+        metric_data_list = TrackingMetricDataList()
+        for class_name in self.cfg.class_names:
+            ev = TrackingEvaluation(self.tracks_gt, self.tracks_pred, class_name, self.cfg.dist_fcn_callable,
+                                    self.cfg.dist_th_tp, self.cfg.min_recall, num_thresholds=TrackingMetricData.nelem,
+                                    verbose=self.verbose)
+            md = ev.accumulate()
+            metric_data_list.set(class_name, md)
+
+        # -----------------------------------
+        # Step 2: Calculate metrics from the data.
```

Comment:
Is this description correct? It looks like we are just reading / aggregating the metrics rather than calculating them.

Reply:
AMOTA/AMOTP are averaged. I guess "aggregate" is a more general term.